Introduce a default synchronization context for wasm

kg opened the pull request:
This PR introduces a JSSynchronizationContext that is installed automatically on the main thread when we initialize the bindings. It implements both Send and Post on top of emscripten's main thread call proxying API when you use it from workers, and on the main thread it just invokes the callback immediately.

I tried to keep the implementation as simple as possible without making it slow enough to produce a meaningful performance regression, since things like await will use this by default once it's merged. The queue of jobs is maintained in managed code and pumped by a dispatcher, and if we notice that the dispatcher is not currently waiting to run when we add something to the queue, we kick off a request to run the dispatcher. This means that worst-case we will run the dispatcher once per work item, but if a bunch of work items end up in the queue, they can get serviced efficiently by a single invocation of the dispatcher. And for calls from the main thread the performance should be equivalent to what it was before due to the fast-path.

lambdageek commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -25,6 +25,12 @@
+#ifndef DISABLE_THREADS
+
+#include <emscripten/threading.h>
+#include <emscripten/wasm_worker.h>
```
We shouldn't be using anything from wasm_worker.h - that's the non-pthreads alternate worker interface that they're developing. It's a completely separate set of workers.

kg replied:
Yeah the docs are really bad so I tried that header first, I think it's from threading.h.

lambdageek commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -471,6 +472,17 @@ export function bindings_lazy_init(): void {
 
     _create_primitive_converters();
 
+    if (!ENVIRONMENT_IS_PTHREAD) {
+        // Install our sync context so that async continuations will migrate back to the main thread automatically
```
Couple of things:

1. I'm not sure if this method still exists after the various startup.ts refactorings.
2. Is this installing a sync context on every non-main pthread?  Why?
3. I think if we want to do work in every non-main pthread, we should be doing it in a new `    pthreads_worker.currentWorkerThreadEvents.addEventListener(pthreads_worker.dotnetPthreadCreated, (event) => ...)` callback.  (see `mono_wasm_pthread_worker_init`) we want to do work every time a pthread is starts (but again... is that what we want?) not every time a new worker is created by Emscripten (and inparticular before the worker has a pthread first assigned to it, it seems too early to be calling wasm code from the worker)

kg replied:
Right now it only installs it on the main thread, and then the concurrent apis like task will capture the sync context and bring it with them to the threadpool threads, I think? I'm not sure whether I like that or not. I'm open to doing it any particular way.

Good point about workers vs pthreads, I'll see if it's reasonable to do it in that worker_init function.

lambdageek replied:
I think I misread what is happening. I saw `if (!ENVIRONMENT_IS_PTHREAD) {` and didn't see the `!`. sorry.

lambdageek commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -1467,4 +1480,16 @@ EMSCRIPTEN_KEEPALIVE int mono_wasm_f64_to_i52 (int64_t *destination, double valu
\ No newline at end of file
+}
+
+EMSCRIPTEN_KEEPALIVE int mono_wasm_dispatch_to_main_thread (int sync, void *callback, void *arg0, void *arg1) {
+#if DISABLE_THREADS
+	return 0;
+#else
+	if (sync)
+		emscripten_sync_run_in_main_runtime_thread(EM_FUNC_SIG_VII, callback, arg0, arg1);
+	else
+		emscripten_async_run_in_main_runtime_thread(EM_FUNC_SIG_VII, callback, arg0, arg1);
+	return 1;
+#endif
+}
```
I can't explain exactly why, but I have a feeling it would be better if this went through the runtime's wrapper around the emscripten primitive (ie call `mono_threads_wasm_async_run_in_main_thread_vii` and also make `mono_threads_wasm_sync_run_in_main_thread_vii` in `mono/utils/mono-threads-wasm.h`).

kg replied:
That sounds fine to me, I'll do it.

lambdageek commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -471,6 +472,17 @@ export function bindings_lazy_init(): void {
+        const sync_context_class = cwraps.mono_wasm_assembly_find_class(
+                binding_module, "System.Runtime.InteropServices.JavaScript", "JSSynchronizationContext"
+            ), sync_context_install_method = cwraps.mono_wasm_assembly_find_method(
+                sync_context_class, "InstallSynchronizationContext", 1
```
Do we have to call `InstallSynchronizationContext` from JS?  Can it be done from [`StartCallback`](https://github.com/dotnet/runtime/blob/028e556926ad4b03afba9ac1d616cfd3538d36cf/src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs#L229-L237) ?

But note StartCallback won't be called for:
1. The main thread
2. external (ie native C) pthreads that attach to the runtime

kg replied:
We need to install it on the main thread. External threads won't matter even if we decide to install it on the managed worker threads, though.

lambdageek replied:
yea I misunderstood what the code was doing.  this seems like a reasonable place to install the sync context, then

lambdageek reviewed (commented):
I took a pass through the concurrent logic and it _seems_ ok (and I will take a more detailed look if we decide to go with it), but I'd like to understand first if it's possible to use some higher-level abstraction instead. We're setting up a channel that workers write into and that the main thread reads from.  Can we use an _actual_ channel?

kg commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs:
```diff
@@ -0,0 +1,150 @@
+
+        private static JSSynchronizationContext? MainThreadSynchronizationContext;
+        private static readonly ConcurrentDictionary<int, byte> PendingWorkItems = new ();
+        private static int PumpIsPending, NextItemId;
```
They can't be volatile, it generates a compiler warning which fails the build. I think all the reads and writes are using Volatile or Interlocked anyway, but I'll double-check.

kg replied:
As for using an event or a semaphore - we never wait on PumpIsPending, it just determines whether we queue a new pump. So I'm not sure there's any value in using a heavier primitive like that, but I'm open to doing it. I guess we could TryEnter it to query its state?

lambdageek replied:
I'm not even particularly invested in any synchronization primitive, if none of them are a good fit.

I think what I'm really after is to isolate the `int` into a helper struct and wrap the `Interlocked`/`Volatile` operations on it in some higher-level functions.  So that it's easier to treat it as some kind of abstraction when reading the code.

kg commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs:
```diff
@@ -0,0 +1,150 @@
+        private static readonly ConcurrentDictionary<int, byte> PendingWorkItems = new ();
+        private static int PumpIsPending, NextItemId;
+
+        private readonly ConcurrentQueue<WorkItem> Queue;
```
I never heard of this type before now, I'll look into it. It sounds like it would work.

lambdageek commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs:
```diff
@@ -0,0 +1,143 @@
+            if (onMainThread)
+                Pump(Queue);
+            else
+                DispatchToMainThread(0, (void*)(delegate* unmanaged[Cdecl]<IntPtr, int, void>)&RequestPumpCallback, IntPtr.Zero, -1);
```
`ScheduleBackgroundJob` (nee `mono_threads_schedule_background_job`) already knows to check if it's on the main thread or a worker.  So you don't need to dispatch calls to it to the main thread.

ie, I think you can just say

```csharp
   if (onMainThread)
        Pump(Queue);
   else
       ScheduleBackgroundJob((void*)(delegate* unmanaged[Cdecl]<void>)&BackgroundJobHandler);
```

lambdageek commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs:
```diff
@@ -0,0 +1,143 @@
+                return;
+            }
+
+            var awaiter = vt.AsTask().ConfigureAwait(false).GetAwaiter();
+            awaiter.UnsafeOnCompleted(_DataIsAvailable);
```
it might be worth a comment to explain what this is doing.

lambdageek reviewed (commented):
The design makes sense to me

Found a few minor things in the implementation, but it's cleanup, not deep issues.

lambdageek commented on src/mono/wasm/runtime/driver.c:
```diff
@@ -135,6 +141,8 @@ typedef struct {
 	const char* (*lookup_icall_symbol) (void* func);
 } MonoIcallTableCallbacks;
 
+EMSCRIPTEN_KEEPALIVE int mono_wasm_dispatch_to_main_thread (int sync, void *callback);
```
This doesn't have an implementation anymore

kg committed:
Introduce a JSSynchronizationContext that automatically remotes function calls back to the browser thread and queues them as background jobs.
Exercise sync context in threads sample to display the complete progress indicator
Undo changes to normal browser sample
Update a comment and address PR feedback
Remove prints and update comments
Cleanup debugging changes
Migrate install method to new marshaling API and clean up merge damage

davidfowl commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs:
```diff
@@ -0,0 +1,140 @@
+
+        public override void Post (SendOrPostCallback d, object? state) {
+            var workItem = new WorkItem(d, state, null);
+            while (!Queue.Writer.TryWrite(workItem))
```
When would this ever return false? The channel is unbounded.

kg replied:
The documentation was unclear on when this could fail, and I didn't have a chance to read it yet since there are multiple queue implementations. Wasn't sure if it was a wait-free thing where TryWrite failing means 'try again, someone had the lock'

davidfowl replied:
cc @IEvangelist a chance to improve the docs! Though this might be an API doc.

pavelsavara commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -578,6 +579,14 @@ export function bindings_init(): void {
 
+        if (!ENVIRONMENT_IS_PTHREAD) {
+            // Install our sync context so that async continuations will migrate back to the main thread automatically
+            const fn = mono_bind_static_method(
```
Please convert this to `[JSExport]` and move it to `init_managed_exports()`

pavelsavara commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs:
```diff
@@ -94,16 +94,27 @@ private JSSynchronizationContext (Thread mainThread)
+        [MethodImpl(MethodImplOptions.NoInlining)] // https://github.com/dotnet/runtime/issues/71425
+        // the marshaled signature is:
+        // void InstallSynchronizationContext()
+        internal static void InstallSynchronizationContext (JSMarshalerArgument* arguments_buffer) {
```
if would be good to move this to `JavaScriptExports.cs` so that we have them all in the same place. That class is also already protected from trimming.

pavelsavara replied:
I mean, the marshaling part of this method, the actual init, could stay here.

pavelsavara commented on src/mono/wasm/runtime/managed-exports.ts:
```diff
@@ -38,6 +23,23 @@ export function init_managed_exports(): void {
     if (!runtimeHelpers.runtime_interop_exports_class)
         throw "Can't find " + runtimeHelpers.runtime_interop_namespace + "." + runtimeHelpers.runtime_interop_exports_classname + " class";
 
+    const install_sync_context = mono_method_resolve(
```
Would this managed code exist in the non-MT build ? 
We have `import MonoWasmThreads from "consts:monoWasmThreads"`
and `if (MonoWasmThreads )  {`

And perhaps some `#if` ON c# side.

kg replied:
The current behavior is that we have the default synchronization context, which sends everything to the threadpool. We want to always install this even if threads are inactive, so that users get consistent behavior.

pavelsavara commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Legacy/Runtime.cs:
```diff
@@ -9,6 +9,11 @@ namespace System.Runtime.InteropServices.JavaScript
     public static class Runtime
     {
+        static unsafe Runtime () {
+            // HACK: For some reason the linker trims this class even if I set all the attributes to disable trimming.
```
with the code in `JavaScriptExports` you probably would not have to do this.

kg replied:
Sadly I still need it. I feel like something is wrong with the linker.

pavelsavara replied:
mark the method `public` because we protect it as `[DynamicDependency(DynamicallyAccessedMemberTypes.PublicMethods`

lambdageek commented on src/mono/wasm/runtime/managed-exports.ts:
```diff
@@ -38,6 +23,23 @@ export function init_managed_exports(): void {
     if (!runtimeHelpers.runtime_interop_exports_class)
         throw "Can't find " + runtimeHelpers.runtime_interop_namespace + "." + runtimeHelpers.runtime_interop_exports_classname + " class";
 
+    const install_sync_context = mono_method_resolve(
+        "[System.Runtime.InteropServices.JavaScript] System.Runtime.InteropServices.JavaScript.JSSynchronizationContext:InstallSynchronizationContext"
+    );
+    mono_assert(install_sync_context, "Can't find InstallSynchronizationContext method");
+    runtimeHelpers.javaScriptExports.install_synchronization_context = () => {
+        const sp = anyModule.stackSave();
+        try {
+            const args = alloc_stack_frame(2);
+            invoke_method_and_handle_exception(install_sync_context, args);
+        } finally {
+            anyModule.stackRestore(sp);
+        }
+    };
```
So basically everything LGTM, except this part, which I don't understand. It seems like you're doing what the JSExport source generator is supposed to do, but by hand? but you're also using `mono_method_resolve` from the legacy API?  I'm not saying it's wrong - just that I don't know how to tell if it's right.  @pavelsavara can you take a look at this bit.

kg replied:
`mono_method_resolve` makes sense to use instead of doing it by hand, even if we're not making it public anymore. I don't see any reason not to use it.

And yeah, I'm duplicating JSExport by hand in this case. Pavel said we do it for a few other things in the runtime and it's the most straightforward option here since we don't need any argument or return value marshaling.

lambdageek replied:
Makes sense, thanks!

pavelsavara replied:
for the internal methods which need to work early during runtime startup, this is good option, to not use the generator etc.

pavelsavara replied:
you should use simpler `get_method()` from `src\mono\wasm\runtime\managed-exports.ts` instead of `mono_method_resolve`.

lambdageek reviewed (approved):
LGTM ~except the managed-exports.ts bits which I don't understand~

lambdageek commented:
/azp run runtime-wasm

kg commented:
Lots of failures on the wasm lanes but none of them look related to this PR. Not sure what to do about that. I see some temp file exists errors that seem to be some sort of problem with the build machine/build environment (I didn't touch native IO at all) and then a bunch of problems in websockets that appear related to SharedArrayBuffer, which I also didn't change.

radical commented:
The `LibraryTests` one is https://github.com/dotnet/runtime/issues/73721 .
The AOT one is "like" https://github.com/dotnet/runtime/issues/72752 - in the sense that it has the same unhelpful, non-specific error message!

kg committed:
Rearrange code to remove linker workaround

pavelsavara commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs:
```diff
@@ -0,0 +1,133 @@
+using System.Runtime.CompilerServices;
+using QueueType = System.Threading.Channels.Channel<System.Runtime.InteropServices.JavaScript.JSSynchronizationContext.WorkItem>;
+
+namespace System.Runtime.InteropServices.JavaScript {
```
Please add short comment on what it does, why and how.

kg committed:
Add additional comments and improve error handling
